### PR TITLE
JAMES-2555 Task wrapping for re-indexing

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+    }
+
+    public Task.Result reIndex(MailboxPath path, ReprocessingContext reprocessingContext) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(path.getUser());
+        return reIndex(path, mailboxSession, reprocessingContext);
+    }
+
+    public Task.Result reIndex(ReprocessingContext reprocessingContext) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing");
```
Comment (rouazana) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: you don't need to create the session of the user?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                    return Task.Result.PARTIAL;
+                }
+            })
+            .reduce(Task::combine)
+            .orElse(Task.Result.COMPLETED);
+    }
+
+    private Task.Result reIndex(MailboxPath path, MailboxSession mailboxSession, ReprocessingContext reprocessingContext) throws MailboxException {
+        MailboxRegistration mailboxRegistration = new MailboxRegistration(path);
+        LOGGER.info("Intend to reindex {}",path);
```
Comment (rouazana) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: missing space

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            .map(globalRegistration::getPathToIndex)
+            .flatMap(OptionalUtils::toStream)
+            .map(path -> {
+                try {
+                    return reIndex(path, reprocessingContext);
+                } catch (Throwable e) {
+                    LOGGER.error("Error while proceeding to full reindexing on {}", path, e);
+                    return Task.Result.PARTIAL;
+                }
+            })
```
Comment (rouazana) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: could be on previous line

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 1 above)
Comment (aduprat) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: This `session` is only used for listing mailboxes, I don't think it has an impact.

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    private Task.Result handleFullReindexingIterations(List<MailboxPath> mailboxPaths, GlobalRegistration globalRegistration,
+                                                       ReprocessingContext reprocessingContext) {
+        return mailboxPaths.stream()
+            .map(globalRegistration::getPathToIndex)
+            .flatMap(OptionalUtils::toStream)
+            .map(path -> {
+                try {
+                    return reIndex(path, reprocessingContext);
+                } catch (Throwable e) {
+                    LOGGER.error("Error while proceeding to full reindexing on {}", path, e);
```
Comment (aduprat) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: `path.asString()`

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 2 above)
Comment (aduprat) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: `path.asString()`

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            return Iterators.toStream(
+                mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
+                    .findInMailbox(mailbox, MessageRange.all(), MessageMapper.FetchType.Metadata, NO_LIMIT))
+                .map(MailboxMessage::getUid)
+                .map(uid -> handleMessageReIndexing(mailboxSession, mailboxRegistration, mailbox, uid))
+                .peek(reprocessingContext::updateAccordingToReprocessingResult)
+                .reduce(Task::combine)
+                .orElse(Task.Result.COMPLETED);
+        } finally {
+            LOGGER.info("Finish to reindex {}", path);
```
Comment (aduprat) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: `path.asString()`

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            Optional<ImpactingMessageEvent> impactingMessageEvent = findMostRelevant(mailboxRegistration.getImpactingEvents(uid));
+
+            Optional.of(uid)
+                .filter(x -> !wasDeleted(impactingMessageEvent))
+                .flatMap(Throwing.function(mUid -> fullyReadMessage(mailboxSession, mailbox, mUid)))
+                .map(message -> messageUpdateRegardingEvents(message, impactingMessageEvent))
+                .ifPresent(Throwing.consumer(message -> messageSearchIndex.add(mailboxSession, mailbox, message)));
+            return Task.Result.COMPLETED;
+        } catch (Exception e) {
+            LOGGER.warn("ReIndexing failed for {} {}", mailbox.generateAssociatedPath(), uid, e);
```
Comment (aduprat) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: Adding an exception in _slf4j_ API is working only if you have at most one other parameter.
`void warn(String msg, Throwable t)`
`void warn(String format, Object arg1, Object arg2)`

**Comment by aduprat** [comment]
```
[ERROR] Errors: 
[ERROR]   ElasticSearchUidSearchOnIndexTest>UidSearchOnIndex.testSearchAtomsUS:52 » InvalidServerResponse
```

**Comment by aduprat** [comment]
test this please

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 6 above)
Comment (rouazana) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: no, I think it's working as expected (confirmed by @jeantil if I remember correctly), that's a hidden feature :)

Comment (aduprat) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: Never use hidden feature, as it may break silently.

**Comment by aduprat** [comment]
```
[ERROR] Errors: 
[ERROR]   RabbitMQClusterTest$ClusterSharing.queuesShouldBeDeclarableOnAnotherNode » Completion
[ERROR]   RabbitMQClusterTest$ClusterSharing.queuesShouldBeShared » Completion org.testc...
[ERROR]   RabbitMQClusterTest$ClusterSharing.rabbitMQManagerShouldReturnThreeNodesWhenAskingForStatus » Completion
```
test this please

**Comment by rouazana** [comment]
```
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 9.587 s <<< FAILURE! - in org.apache.james.app.spring.JamesAppSpringMainTest
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR] testServer(org.apache.james.app.spring.JamesAppSpringMainTest)  Time elapsed: 9.533 s  <<< ERROR!
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'exporter' defined in class path resource [META-INF/org/apache/james/spring-server.xml]: Cannot resolve reference to bean 'reindexermanagement' while setting bean property 'beans' with key [TypedStringValue: value [org.apache.james:type=component,name=reindexerbean], target type [null]]; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'reindexermanagement' defined in class path resource [META-INF/org/apache/james/spring-server.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [org.apache.james.adapter.mailbox.ReIndexerManagement]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.apache.james.adapter.mailbox.ReIndexerManagement.<init>()
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'reindexermanagement' defined in class path resource [META-INF/org/apache/james/spring-server.xml]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [org.apache.james.adapter.mailbox.ReIndexerManagement]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.apache.james.adapter.mailbox.ReIndexerManagement.<init>()
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] Caused by: org.springframework.beans.BeanInstantiationException: Could not instantiate bean class [org.apache.james.adapter.mailbox.ReIndexerManagement]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.apache.james.adapter.mailbox.ReIndexerManagement.<init>()
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] Caused by: java.lang.NoSuchMethodException: org.apache.james.adapter.mailbox.ReIndexerManagement.<init>()
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] 
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [INFO] 
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [INFO] Results:
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [INFO] 
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR] Errors: 
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR]   JamesAppSpringMainTest.testServer:27 » BeanCreation Error creating bean with n...
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR] org.apache.james.app.spring.JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener(org.apache.james.app.spring.JamesSpringContextTest)
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR]   Run 1: JamesSpringContextTest.setup:45 » BeanCreation Error creating bean with name '...
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR]   Run 2: JamesSpringContextTest.tearDown:52 NullPointer
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [INFO] 
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [INFO] 
[7a29f8005c38ccc8573e6862ccd05c77f0570dd1] [ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0
```

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 6 above)
Comment (jeantil) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: the alternative is highly verbose :/

**Comment by aduprat** [comment]
I think you've break something
```
[ERROR] Errors: 
[ERROR]   ReindexCommandIntegrationTest.reindexMailboxShouldWork » ThrownByLambda java.u...
```
maybe the injection is the root cause.

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 3 above)
Comment (rouazana) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: agree, and checkstyle should prevent to do that (with a so niiiiiice array out of bound error)

**Comment by chibenwa** [comment]
```
[2aa16a0dd6c15cb5e2611d42d33c328f003335dd] fatal: Cannot update paths and switch to branch 'null' at the same time.
[2aa16a0dd6c15cb5e2611d42d33c328f003335dd] Did you intend to checkout 'apache/null' which can not be resolved as commit?
[2aa16a0dd6c15cb5e2611d42d33c328f003335dd] Running: Shell Script
```

**Comment by chibenwa** [comment]
test this please

CI should be fixed

**Comment by chibenwa** [comment]
```

[2aa16a0dd6c15cb5e2611d42d33c328f003335dd] [ERROR] Errors: 
[2aa16a0dd6c15cb5e2611d42d33c328f003335dd] [ERROR]   ElasticSearchUidSearchOnIndexTest>UidSearchOnIndex.testSearchAtomsUS:52 » InvalidServerResponse
```

test this please

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        LOGGER.info("Intend to reindex {}", path.asString());
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
+        messageSearchIndex.deleteAll(mailboxSession, mailbox);
+        mailboxManager.addListener(path, mailboxRegistration, mailboxSession);
+        try {
+            return Iterators.toStream(
+                mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
+                    .findInMailbox(mailbox, MessageRange.all(), MessageMapper.FetchType.Metadata, NO_LIMIT))
+                .map(MailboxMessage::getUid)
+                .map(uid -> handleMessageReIndexing(mailboxSession, mailboxRegistration, mailbox, uid))
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: Can we use FluentFutureStream in here? I see we are processing message sequenitally

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+import org.apache.mailbox.tools.indexer.events.ImpactingMessageEvent;
+import org.apache.mailbox.tools.indexer.registrations.GlobalRegistration;
+import org.apache.mailbox.tools.indexer.registrations.MailboxRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.Lists;
+
+public class ReIndexerPerformer {
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: default modifier?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+    private static final int NO_LIMIT = 0;
+    private static final int SINGLE_MESSAGE = 1;
+
+    private final MailboxManager mailboxManager;
+    private final ListeningMessageSearchIndex messageSearchIndex;
+    private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
+
+    @Inject
+    public ReIndexerPerformer(MailboxManager mailboxManager,
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: default modifier?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+                              ListeningMessageSearchIndex messageSearchIndex,
+                              MailboxSessionMapperFactory mailboxSessionMapperFactory) {
+        this.mailboxManager = mailboxManager;
+        this.messageSearchIndex = messageSearchIndex;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+    }
+
+    public Task.Result reIndex(MailboxPath path, ReprocessingContext reprocessingContext) throws MailboxException {
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: default modifier?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(path.getUser());
+        return reIndex(path, mailboxSession, reprocessingContext);
+    }
+
+    public Task.Result reIndex(ReprocessingContext reprocessingContext) throws MailboxException {
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: default modifier?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/FullReindexingTask.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.task.Task;
+import org.apache.james.task.TaskExecutionDetails;
+
+public class FullReindexingTask implements Task {
+
+    public static final String FULL_RE_INDEXING = "FullReIndexing";
+
+    public static class AdditionalInformation implements TaskExecutionDetails.AdditionalInformation {
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/FullReindexingTask.java]: `AdditionalInformation` is used only inside this class, private qualifier?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTask.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.task.Task;
+import org.apache.james.task.TaskExecutionDetails;
+
+public class SingleMailboxReindexingTask implements Task {
+
+    public static final String MAILBOX_RE_INDEXING = "mailboxReIndexing";
+
+    public static class AdditionalInformation implements TaskExecutionDetails.AdditionalInformation {
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTask.java]: `AdditionalInformation` is used only inside this class, private qualifier?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: Well : 
 - the mailbox APIs do not work with futures so we would not gain anything
 - From my experience fetching multiple messagesin parallel is a bit too heavy for Cassandra. You have to proceed with batches. WHich brings additional complexity anyway. 

Hence I propose to leave with current (safe) sequential code and come back later if optimisation is needed.

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/FullReindexingTask.java (hunk 12 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/FullReindexingTask.java]: It is used to transmit information about this class to the outer world.

JSON serialization reflexion allows to duisplay task details as part of our APIs (better design we came up with for it).

Hence a public qualifier makes it clear this class is exposed and used by the outer world, not just inside this class.

**Comment by chibenwa** [comment]
Merged

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java (hunk 6 above)
Comment (mbaechler) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: It's not hidden, just not type-checked

**Pull request closed without merging**